GVFS/GVFS.Common/Maintenance/PackfileMaintenanceStep.cs

```diff
@@ -0,0 +1,122 @@
+﻿using GVFS.Common.FileSystem;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace GVFS.Common.Maintenance
+{
+    public class PackfileMaintenanceStep : GitMaintenanceStep
+    {
+        public const string PackfileLastRunFileName = "pack-maintenance.time";
+        public const string DefaultBatchSize = "2g";
+        private readonly bool forceRun;
+        private readonly string batchSize;
+
+        public PackfileMaintenanceStep(
+            GVFSContext context,
+            bool requireObjectCacheLock = true,
+            bool forceRun = false,
+            string batchSize = DefaultBatchSize)
+            : base(context, requireObjectCacheLock)
+        {
+            this.forceRun = forceRun;
+            this.batchSize = batchSize;
+        }
+
+        public override string Area => nameof(PackfileMaintenanceStep);
+        protected override string LastRunTimeFilePath => Path.Combine(this.Context.Enlistment.GitObjectsRoot, "info", PackfileLastRunFileName);
+        protected override TimeSpan TimeBetweenRuns => TimeSpan.FromDays(1);
+
+        // public only for unit tests
+        public void CountPackFiles(out int count, out long size)
+        {
+            count = 0;
+            size = 0;
+
+            foreach (DirectoryItemInfo info in this.Context.FileSystem.ItemsInDirectory(this.Context.Enlistment.GitPackRoot))
+            {
+                if (string.Equals(Path.GetExtension(info.Name), ".pack", StringComparison.OrdinalIgnoreCase))
+                {
+                    count++;
+                    size += info.Length;
+                }
+            }
+        }
+
+        // public only for unit tests
+        public IEnumerable<string> CleanStaleIdxFiles()
+        {
+            List<DirectoryItemInfo> packDirContents = this.Context
+                                                          .FileSystem
+                                                          .ItemsInDirectory(this.Context.Enlistment.GitPackRoot)
+                                                          .ToList();
+
+            foreach (DirectoryItemInfo info in packDirContents)
+            {
+                if (string.Equals(Path.GetExtension(info.Name), ".idx", StringComparison.OrdinalIgnoreCase))
+                {
+                    string pairedPack = info.FullName.Substring(0, info.FullName.Length - "idx".Length) + "pack";
+
+                    if (!this.Context.FileSystem.FileExists(pairedPack))
+                    {
+                        if (this.Context.FileSystem.TryDeleteFile(info.FullName))

```

Comment:
If we can't delete the file should we log a warning?

Reply:
We log the number of stale IDX files in the metadata below, which is tracked by telemetry.

Reply:
Could this fail due to a something in GVFS having a handle to it?  Reading your explaination on the PR.  If there are expected fail cases it would be helpful to add comments here.